File: ExtractCardInfo.java
package name.frw;

import java.awt.GridLayout;
import java.awt.event.ActionEvent;
import java.awt.event.ActionListener;
import java.io.File;
import java.io.PrintWriter;
import java.io.StringReader;
import java.io.StringWriter;

import javax.swing.JButton;
import javax.swing.JFrame;
import javax.swing.JLabel;
import javax.swing.JOptionPane;
import javax.swing.JRadioButton;
import javax.swing.JTextField;
import javax.swing.SwingUtilities;
import javax.xml.parsers.DocumentBuilderFactory;
import javax.xml.transform.OutputKeys;
import javax.xml.transform.Transformer;
import javax.xml.transform.TransformerException;
import javax.xml.transform.TransformerFactory;
import javax.xml.transform.dom.DOMSource;
import javax.xml.transform.stream.StreamResult;
import javax.xml.xpath.XPath;
import javax.xml.xpath.XPathConstants;
import javax.xml.xpath.XPathExpression;
import javax.xml.xpath.XPathFactory;

import org.w3c.dom.Document;
import org.w3c.dom.Element;
import org.w3c.dom.Node;
import org.w3c.dom.NodeList;
import org.xml.sax.InputSource;

public class ExtractCardInfo extends JFrame {
	private static final long serialVersionUID = -7844908032131837091L;
	private static final String VERSION_NO = "5";
	private static final int HEIGHT_OF_WINDOW = 250;
	
	private final JTextField fileField = new JTextField("<folder path>\\CardDefs.xml");
	private final JRadioButton debugRButton = new JRadioButton("Debug mode", false);
	private final JLabel resultLabel = new JLabel("You need to download CardDefs.xml from https://github.com/HearthSim/hs-data");
	private final JButton startButton = new JButton("Start");
	
	private static ExtractCardInfo tm;

	private static PrintWriter pw;

	private ExtractCardInfo() {
		super("Extract Card Information from hs-data" + " v" + VERSION_NO);
		startButton.addActionListener(new ActionListener() {
			public void actionPerformed(ActionEvent e) {
				String sourceFile = fileField.getText();

				try {
					if (sourceFile.equals("")) {
						throw new Exception("blank file path.");
					}

					pw = new PrintWriter("cardDB.txt");
//					pw.println("Cost\tName\tClass\tRarity\tCardSet\tCardID\tCardType");
					pw.println("Cost\tName\tClass\tRarity\tCardSet\tCardType");
					extract(sourceFile);
					pw.close();

					JOptionPane.showMessageDialog(null, "Done.");
				} catch (Exception exception) {
					JOptionPane.showMessageDialog(null, "Error happens:\n" + exception.getMessage());
				}
			}

			private String nodeToString(Node node) {
				StringWriter sw = new StringWriter();
				try {
					Transformer t = TransformerFactory.newInstance().newTransformer();
					t.setOutputProperty(OutputKeys.OMIT_XML_DECLARATION, "yes");
					t.transform(new DOMSource(node), new StreamResult(sw));
				} catch (TransformerException te) {
					JOptionPane.showMessageDialog(null, "Error happens:\n" + te.getMessage());
				}
				return sw.toString();
			}

			private String changeValue2Class(String value) {
				String text = "";

				if (value.equals("2")) {
					text = "Druid";
				} else if (value.equals("3")) {
					text = "Hunter";
				} else if (value.equals("4")) {
					text = "Mage";
				} else if (value.equals("5")) {
					text = "Paladin";
				} else if (value.equals("6")) {
					text = "Priest";
				} else if (value.equals("7")) {
					text = "Rogue";
				} else if (value.equals("8")) {
					text = "Shaman";
				} else if (value.equals("9")) {
					text = "Warlock";
				} else if (value.equals("10")) {
					text = "Warrior";
				} else if (value.equals("12")) {
					text = "Neutral";
				}

				return checkDebugMode(value, text);
			}

			private String changeValue2Rarity(String value, String cardSet) {
				if (cardSet.equals("2")) {
					return checkDebugMode(value, "Basic");
				}
				
				String text = "";

				if (value.equals("1")) {
					text = "Common";
				} else if (value.equals("2")) {
					text = "Basic";
				} else if (value.equals("3")) {
					text = "Rare";
				} else if (value.equals("4")) {
					text = "Epic";
				} else if (value.equals("5")) {
					text = "Legendary";
				}

				return checkDebugMode(value, text);
			}

			private String changeValue2CardType(String value) {
				String text = "";

				if (value.equals("4")) {
					text = "Minion";
				} else if (value.equals("5")) {
					text = "Spell";
				} else if (value.equals("7")) {
					text = "Weapon";
				} else if (value.equals("3")) {
					text = "Hero";
				}

				return checkDebugMode(value, text);
			}

			private String checkDebugMode(String value, String text) {
				if (debugRButton.isSelected()) {
					return value + text;
				}
				return text;
			}

			private void extract(String sourceFile) throws Exception {
				Document doc = DocumentBuilderFactory.newInstance().newDocumentBuilder().parse(new File(sourceFile));
				XPath xpath = XPathFactory.newInstance().newXPath();
				XPathExpression expr = xpath.compile("/CardDefs/Entity[Tag[@enumID='321' and @value='1']]");
				NodeList nodes = (NodeList) expr.evaluate(doc, XPathConstants.NODESET);
				for (int i = 0; i < nodes.getLength(); i++) {
					Element entity = (Element) nodes.item(i);
					doc = DocumentBuilderFactory.newInstance().newDocumentBuilder().parse(new InputSource(new StringReader(nodeToString(entity))));

					String cardCost = xpath.compile("/Entity/Tag[@enumID='48']/@value").evaluate(doc);
					if (cardCost.equals("")) {
						cardCost = "0";
					}
					String cardName = xpath.compile("/Entity/Tag[@enumID='185']/enUS").evaluate(doc);
					String cardClass = xpath.compile("/Entity/Tag[@enumID='199']/@value").evaluate(doc);
					String cardCardType = xpath.compile("/Entity/Tag[@enumID='202']/@value").evaluate(doc);
					String cardRarity = xpath.compile("/Entity/Tag[@enumID='203']/@value").evaluate(doc);
					String cardCardSet = xpath.compile("/Entity/Tag[@enumID='183']/@value").evaluate(doc);
//					String cardID = entity.getAttribute("CardID");

//					String text = cardCost + "\t" + cardName + "\t" + changeValue2Class(cardClass) + "\t" + changeValue2Rarity(cardRarity, cardCardSet) + "\t" + cardCardSet + "\t" + cardID + "\t" + changeValue2CardType(cardCardType);
					String text = cardCost + "\t" + cardName + "\t" + changeValue2Class(cardClass) + "\t" + changeValue2Rarity(cardRarity, cardCardSet) + "\t" + cardCardSet + "\t" + changeValue2CardType(cardCardType);
					pw.println(text);
				}
			}
		});

		setLayout(new GridLayout(0, 1));
		add(new JLabel("File path:"));
		add(fileField);
		add(debugRButton);

		add(resultLabel);
		add(startButton);
	}

	public static void main(String[] args) {
		SwingUtilities.invokeLater(new Runnable() {
			public void run() {
				tm = new ExtractCardInfo();
				tm.setDefaultCloseOperation(JFrame.EXIT_ON_CLOSE);
				tm.setSize(600, HEIGHT_OF_WINDOW);
				tm.setLocationRelativeTo(null);
				tm.setVisible(true);
			}
		});
	}
}

File: ExtractCardInfo.pl
#!/usr/bin/perl
use warnings;
use strict;

use XML::XPath;
use XML::XPath::XMLParser;


#1. Get <Entity>
#2. Parse <Entity>
#	2.1. Collectible?		<Tag enumID="321" name="Collectible" type="Bool" value="1">
#		2.1.1. Next unless value is true
#   2.2. Hero?				<Tag enumID="202" name="CardType" type="CardType" value="3"/>
#       2.2.1. Next if value == 3
#   2.2. We need to get:
#		2.2.1. Cost			<Tag enumID="48" name="Cost" type="Int" value="5"/>
#		2.2.2. Name			<Tag enumID="185" name="CardName" type="LocString"><enUS>Flame Lance</enUS></Tag>
#		2.2.3. Class		<Tag enumID="199" name="Class" type="Class" value="4"/>
#		2.2.4. Rarity		<Tag enumID="203" name="Rarity" type="Rarity" value="1"/>
#		2.2.5. CardSet		<Tag enumID="183" name="CardSet" type="CardSet" value="15"/>
#		2.2.6. CardID		<Entity CardID="AT_001" version="2">
#		2.2.7. CardType		<Tag enumID="202" name="CardType" type="CardType" value="5"/>
#3. Save to CardDB.txt

#1. Get <Entity>
my $entities_ref = getEntities();
#2. Parse <Entity>
$entities_ref = keepCollectible($entities_ref);
#3. Save to CardDB.txt
saveToTxt($entities_ref);
print "Done\n";

sub saveToTxt {
	#title of each column
	open FH, "> CardDB.txt" or die $!;
	print FH "Cost\tName\tClass\tRarity\tCardSet\tCardID\tCardType\n";
	# 2.1.6. save the temp file
	my $cards_ref = shift;
	for my $card_ref (@{$cards_ref}) {
		print FH $card_ref->{"Cost"},		"\t";
		print FH $card_ref->{"CardName"},	"\t";
		print FH $card_ref->{"Class"},		"\t";
		print FH $card_ref->{"Rarity"},		"\t";
		print FH $card_ref->{"CardSet"},	"\t";
		print FH $card_ref->{"CardID"},		"\t";
		print FH $card_ref->{"CardType"},	"\n";
	}
	close FH;
}

sub keepCollectible {
	my $entities_ref = shift;
	my @entities;
	
	for my $anEntity (@{$entities_ref}) {
		my $xp = XML::XPath->new(xml => $anEntity);
		#2.1. Collectible?
		next unless $xp->findvalue('/Entity/Tag[enumID="321"]/@value');
		#2.2. Hero?
		my $CardType = $xp->findvalue('/Entity/Tag[enumID="202"]/@value');
		next if $CardType == 3;
		#2.3. We need to get:
		my %card = (
			Cost		=> $xp->findvalue('/Entity/Tag[enumID="48"]/@value'),
			CardName	=> $xp->findvalue('/Entity/Tag[enumID="185"]/enUS'),
			Class		=> changeValue2Class($xp->findvalue('/Entity/Tag[enumID="199"]/@value')),
			Rarity		=> changeValue2Rarity($xp->findvalue('/Entity/Tag[enumID="203"]/@value')),
			CardSet		=> $xp->findvalue('/Entity/Tag[enumID="183"]/@value'),
			CardID		=> $xp->findvalue('/Entity/@CardID'),
			CardType	=> changeValue2CardType($CardType)
		);
		push @entities, \%card;
	}

	return \@entities;
}

sub changeValue2Class {
	my $value = shift;
	my $text;
	
	if		($value == 2) { $text = "Druid";}
	elsif	($value == 3) { $text = "Hunter";}
	elsif	($value == 4) { $text = "Mage";}
	elsif	($value == 5) { $text = "Paladin";}
	elsif	($value == 6) { $text = "Priest";}
	elsif	($value == 7) { $text = "Rogue";}
	elsif	($value == 8) { $text = "Shaman";}
	elsif	($value == 9) { $text = "Warlock";}
	elsif	($value == 10){ $text = "Warrior";}
	elsif	($value == 12){ $text = "Neutral";}
	
	return checkDebugMode($value, $text);
}

sub changeValue2Rarity {
	my $value = shift;
	my $text;
	
	if		($value == 1) { $text = "Common";}
	elsif	($value == 2) { $text = "Basic";}
	elsif	($value == 3) { $text = "Rare";}
	elsif	($value == 4) { $text = "Epic";}
	elsif	($value == 5) { $text = "Legendary";}
	
	return checkDebugMode($value, $text);
}

sub changeValue2CardType {
	my $value = shift;
	my $text;
	
	if		($value == 4) { $text = "Minion";}
	elsif	($value == 5) { $text = "Spell";}
	elsif	($value == 7) { $text = "Weapon";}
	elsif	($value == 3) { $text = "Hero";}
	
	return checkDebugMode($value, $text);
}

sub checkDebugMode {
	my $value = shift;
	my $text = shift;
	
	return "$value$text";
}

sub getEntities {
	my @entities;
	my $xp = XML::XPath->new(filename => $ARGV[0]);
	my $nodeset = $xp->find("/CardDefs/Entity");
 	for my $nodes ($nodeset->get_nodelist) {
		my $node = XML::XPath::XMLParser::as_string($nodes);
		push @entities, $node;
	}
	return \@entities;
}
